src/renderer/renderer.cpp

```diff
@@ -0,0 +1,173 @@
+#include "../main.h"
+#include "../util/log.h"
+#include "opengl.h"
+#include "renderer.h"
+
+#include <fstream>
+
+static char *dump_file(const char *file_path) {
+    // TODO(ed): Error handling.
+    std::ifstream in(file_path, std::ios_base::ate);
+    if (in) {
+        auto size = in.tellg();
+        char *buffer = new char [static_cast<u32>(size) + 1];
+        buffer[size] = 0;
+        in.seekg(std::ios_base::beg);
+        in.read(buffer, size);
+        in.close();
+        return buffer;
+    }
+    ERROR("Failed to read \"%s\"", file_path);
+    return nullptr;
+}
+
+namespace GFX {
+
+Shader shader;
+
+void Shader::use() { GL::UseProgram(program_id); }
+
+Mesh Mesh::init(std::vector<Vec3> points) {
+    u32 vao, vbo;
+
+    GL::GenVertexArrays(1, &vao);
+    GL::GenBuffers(1, &vbo);
+
+    GL::BindVertexArray(vao);
+
+    GL::BindBuffer(GL::cARRAY_BUFFER, vbo);
+    GL::BufferData(GL::cARRAY_BUFFER, sizeof(Vec3) * points.size(), &points[0], GL::cSTATIC_DRAW);
+
+    GL::EnableVertexAttribArray(0);
+    GL::VertexAttribPointer(0, 3, GL::cFLOAT, 0, sizeof(Vec3), (void *) 0);
+
+    return {vao, vbo, static_cast<u32>(points.size())};
+}
+
+void Mesh::draw() {
+    GL::BindVertexArray(vao);
+    GL::DrawArrays(GL::cTRIANGLES, 0, draw_length);
+    GL::BindVertexArray(0);
+}
+
+Shader default_shader() {
+    return shader;
+}
+
+Shader compile_shader(const char *source) {
+    auto shader_error_check = [](u32 shader) -> bool {
+        i32 success;
+        GL::GetShaderiv(shader, GL::cCOMPILE_STATUS, &success);
+        if (!success) {
+            char info_log[512];
+            GL::GetShaderInfoLog(shader, sizeof(info_log), NULL, info_log);
```

Comment:
You can get the length of the log with an equivalent call to
```
int max_length = 0;
glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &max_length);
```
and dynamically allocate that length to ensure that the log isn't cut off.

Reply:
I know, but that requires dynamic memory allocations, and more work for very little gain. That code _shouldn't_ have to run, so I don't see the value in the tradeoff since it's 5 extra lines of code for what I see as practically no gain... But I am open for discussion.

Reply:
Probably not worth it then. If it errors and it's cut off we can just increase the `512`.

Reply:
Or fix the error that hopefully takes less than 3 tweets to communicate.